storage/operation/events.go

```diff
@@ -9,11 +13,88 @@ func eventPrefix(prefix byte, blockID flow.Identifier, event flow.Event) []byte
+		return fmt.Errorf("InsertBlockEvents requires LockInsertEvent to be held")
+	}
+
+	// Check if events for the block already exist
+	// We can exit early if we find one existing event E, assuming that the process which wrote E in the past
+	// correctly inserted all other events for the block containing E.
+	// This function only inserts new events; it does not sanity check existing events or ever overwrite events.
+	prefix := MakePrefix(codeEvent, blockID)
+	checkExists := func(key []byte) error {
+		return fmt.Errorf("event with key %x already exists under prefix %x: %w", key, prefix, storage.ErrAlreadyExists)
+	}
+
+	err := IterateKeysByPrefixRange(rw.GlobalReader(), prefix, prefix, checkExists)
+	if err != nil {
+		return err
+	}
+
+	writer := rw.Writer()
+
+	for _, eventsList := range events {
+		for _, event := range eventsList {
+			err := insertEvent(writer, blockID, event)
+			if err != nil {
+				return fmt.Errorf("cannot batch insert event: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// insertEvent stores a regular event in the database.
+// The event is stored with a key that includes the block ID, transaction ID, transaction index, and event index.
+func insertEvent(w storage.Writer, blockID flow.Identifier, event flow.Event) error {
 	return UpsertByKey(w, eventPrefix(codeEvent, blockID, event), event)
 }
 
-func InsertServiceEvent(w storage.Writer, blockID flow.Identifier, event flow.Event) error {
+// InsertBlockServiceEvents stores all service events for a given block in the database.
+// Requires LockInsertServiceEvent to be held for thread safety.
+// This function iterates through all service events in the provided list and stores each event individually.
+// Service events are special events generated by the system (e.g., account creation, contract deployment).
+// It returns [storage.ErrAlreadyExists] if events for the block already exist.
+func InsertBlockServiceEvents(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, events []flow.Event) error {
+	if !lctx.HoldsLock(storage.LockInsertServiceEvent) {
+		return fmt.Errorf("InsertBlockServiceEvents requires LockInsertServiceEvent to be held")
+	}
+
+	// Check if events for the block already exist
+	// We can exit early if we find one existing event E, assuming that the process which wrote E in the past
+	// correctly inserted all other events for the block containing E.
+	// This function only inserts new events; it does not sanity check existing events or ever overwrite events.
+	prefix := MakePrefix(codeServiceEvent, blockID)
+	checkExists := func(key []byte) error {
+		return fmt.Errorf("event with key %x already exists under prefix %x: %w", key, prefix, storage.ErrAlreadyExists)
+	}
+
+	err := IterateKeysByPrefixRange(rw.GlobalReader(), prefix, prefix, checkExists)
+	if err != nil {
+		return err
+	}
+
+	writer := rw.Writer()
+
+	for _, event := range events {
+		err := insertServiceEvent(writer, blockID, event)
```

Comment:
would suggest to inline the body of `insertServiceEvent` here. The documentation of how the key is generated could be moved to the shared method `eventPrefix`